aiohttp/client.py

```diff
@@ -532,6 +532,8 @@ async def _request(
                     except ClientError:
                         raise
                     except OSError as exc:
+                        if exc.errno is None and isinstance(exc, asyncio.TimeoutError):
```

Comment:
Could you explain the `is None` check? Is it just a performance optimization? Why'd you drop the separate block idea?

Reply:
So this is for the case when you get a real OSError with an errno that's also a TimeoutError

I thought about checking for `type(TimeoutError.__cause__) is CancelledError` but your timeout context manager doesn't set the cause and neither does _chain_future

If you do:

```
a, b = socket.socketpair()
with a, b:
    a.settimeout(0.1)
    a.recv(1)
```

You still get a TimeoutError with a None errno so there's no way to actually tell if you have a real operating system TimeoutError or a synthetic one from asyncio

This is why trio uses `trio.TooSlowError` so it's a bit more clear

Reply:
Maybe worth checking if `asyncio.timeout()` has the same problem. We could start migrating from async-timeout to that in Python 3.11+.
https://docs.python.org/3.11/library/asyncio-task.html#timeouts

Reply:
Would be good if we can all agree on some

```
class TooSlowError(asyncio.TimeoutError):
   ...
```

Reply:
Ah, it still uses the same logic from async-timeout:
https://github.com/python/cpython/blob/main/Lib/asyncio/timeouts.py#L98

You'd have to take that up with the cpython issue tracker, or @asvetlov.

Reply:
https://github.com/python/cpython/pull/95830

Reply:
So, if that gets merged, what does the code here become?

Reply:
Fixing _chain_future needs to happen first, but we can check the `__cause__` to see if it's an asyncio Timeout rather than an OS timeout

Reply:
Just coming back round to this, I see the change is now in Python 3.12+. Can we update these checks with a sys.version_info check, so we don't forget about it completely?